tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -594,9 +604,20 @@ internal static CreateAsyncEndMethodResult CreateAsyncEndMethodDelegate(Type int
                 return default;
             }
 
+            bool isTaskReturn = false;
+            Type dynMethodReturnType = returnType;
             if (!onAsyncMethodEndMethodInfo.ReturnType.IsGenericParameter && onAsyncMethodEndMethodInfo.ReturnType != returnType)
             {
-                ThrowHelper.ThrowArgumentException($"The return type of the method: {EndAsyncMethodName} in type: {integrationType.FullName} is not {returnType}");
+                if (onAsyncMethodEndMethodInfo.ReturnType.GetGenericTypeDefinition() == typeof(Task<>) ||
+                    onAsyncMethodEndMethodInfo.ReturnType == typeof(Task))
+                {
+                    dynMethodReturnType = typeof(Task<>).MakeGenericType(returnType);
+                    isTaskReturn = true;
+                }
```

Comment:
I was trying to following the behaviour when the method we're instrumenting returns `Task`. For example

```csharp
Task SomeLibraryFunctionAsync();
```

AFAICT, we have only ever used the format:

```csharp
TReturn OnAsyncMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, CallTargetState state);
```

Which AFAICT means the `TReturn` would be a `Task`? If we need to await this (and I'm understanding correctly) then the simple approach is just to change the return value:

```csharp
async Task<TReturn> OnAsyncMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, CallTargetState state);
```

In this case, `Task<TReturn>` is `Task<Task>`? Will this line ever apply:

```csharp
onAsyncMethodEndMethodInfo.ReturnType == typeof(Task))
```

If so, maybe we should add that to the examples?

Reply:
In `OnAsyncMethodEnd` TReturn is the type returned by the `Task<>` .

For example:

```csharp
Task<string> GetStringAsync()
```
- For `OnMethodEnd` `TReturn` is the actual return type `Task<string>` 
- For `OnAsyncMethodEnd` because we are already inside the async context, `TReturn` is `string`.

Reply:
That's also documented here: https://github.com/DataDog/dd-trace-dotnet/pull/3555/files/4474c007a2229b44ed906069073c4bc1217bd89f#diff-b8474926d22c08287d9b581d5274152d706f2cf6be60127bf7fa3ba455aface5R594-R596

Reply:
> That's also documented here

Well, that's embarrassing 😄 thanks!